Fix the group selector drop down

sharidas opened the pull request:
This change gets the group selector drop
down

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
Fix the group selector in the core. After moving the code to user_management app, the ajax request to get the groups available for the user did not worked. The reason for this is there is no controller in the core to address the request. This PR fixes the issue.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes https://github.com/owncloud/core/issues/31739

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Fix the group selector in the core. After moving the code to user_management app the ajax request to get the groups available for user did not worked in the master branch of core. There was no route and controller to address the request. This PR tries to fix the issue by adding a new route and controller to core which accepts the request and provides the response back.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Login as admin and create users `group1`, `group2`
- Tick the exclude shares check box mentioned in the issue
- Admin user can see `admin`, `group1` and `group2` listed.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

sharidas commented:
Current state of the PR:
- [x] Added `MetaData.php` from `user_management` to core
- [x] Created a new controller for the group selector request.
- [ ] Update the unit tests

DeepDiver1975 commented on settings/Controller/GroupsSelectController.php:
```diff
@@ -0,0 +1,98 @@
+	 * @param int $sortGroups
+	 * @return DataResponse
+	 */
+	public function index($pattern = '', $filterGroups = false, $sortGroups = MetaData::SORT_USERCOUNT) {
```
you need all parameters here?

DeepDiver1975 commented on lib/private/MetaData.php:
```diff
@@ -0,0 +1,200 @@
+
+use OCP\IUserSession;
+
+class MetaData {
```
As explained I don't want to see this class in code

sharidas committed:
Fix the group selector drop down

Fix the group selector drop down

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Current state of the PR:
- Added separate controller for group selector
- This PR has no dependency with user_management. Meaning, user_management remains untouched.
- Added unit test for the change made.

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+				[
+					'status' => 'error',
+					'data' => [
+						'message' => 'No groups available'
```
this message is wrong

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+	 * @param $isAdmin
+	 * @param IUser $user
+	 * @param string $search
+	 * @return array|\OCP\IGroup[]
```
why array? IGroup[] is sufficient

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+use OCP\IUser;
+use OCP\IUserSession;
+
+class GroupsController extends Controller {
```
PHPDoc

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+use OCP\IUserSession;
+
+class GroupsController extends Controller {
+	private $groupManager;
```
PHPDoc

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+class GroupsController extends Controller {
+	private $groupManager;
+	private $userSession;
+	public function __construct(string $appName,
```
PHPDoc

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+	/**
+	 * Get the groups for the user
+	 *
+	 * @NoAdminRequired
```
Usually we add a no csrf annotation to GET requests ... please have a second look and see if needed ...

DeepDiver1975 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,102 @@
+		foreach ($groups as $group) {
+			if ($group->getGID() === 'admin') {
+				$adminGroup['id'] = $group->getGID();
+				$adminGroup['name'] = $group->getGID();
```
there is a getDisplayName afaik - please adapt

DeepDiver1975 commented on tests/Settings/Controller/GroupsControllerTest.php:
```diff
@@ -0,0 +1,142 @@
+use OCP\IUser;
+use OCP\IUserSession;
+
+class GroupsControllerTest extends TestCase {
```
PHPDoc

DeepDiver1975 commented on tests/Settings/Controller/GroupsControllerTest.php:
```diff
@@ -0,0 +1,142 @@
+use OCP\IUserSession;
+
+class GroupsControllerTest extends TestCase {
+	private $appName;
```
PHPDoc

DeepDiver1975 commented on tests/Settings/Controller/GroupsControllerTest.php:
```diff
@@ -0,0 +1,142 @@
+
+class GroupsControllerTest extends TestCase {
+	private $appName;
+	private $request;
```
PHPDoc

DeepDiver1975 commented on tests/Settings/Controller/GroupsControllerTest.php:
```diff
@@ -0,0 +1,142 @@
+class GroupsControllerTest extends TestCase {
+	private $appName;
+	private $request;
+	private $groupManager;
```
PHPDoc

DeepDiver1975 commented on tests/Settings/Controller/GroupsControllerTest.php:
```diff
@@ -0,0 +1,142 @@
+	private $appName;
+	private $request;
+	private $groupManager;
+	private $userSession;
```
PHPDoc

DeepDiver1975 commented on tests/Settings/Controller/GroupsControllerTest.php:
```diff
@@ -0,0 +1,142 @@
+	private $request;
+	private $groupManager;
+	private $userSession;
+	private $groupsController;
```
PHPDoc

sharidas commented:
@DeepDiver1975 kindly review my change. The updated PR includes:
- [x] Updated the message when the user is null
- [x] Added missing PHP Doc ( for both controller and the test )
- [x] Updated the get request with `NoCSRFRequired` annotation. Grepped through the source code, and figured out that the GET requests are using the same. Hence adapting it.
- [x] Updated the code to use `getDisplayName` for the group name ( earlier getGID was used ).
- [x] Updated the code to return `IGroup[]` ( earlier it was array|IGroup[] )

PVince81 commented:
@sharidas mind adding a summary at the top to explain why this PR fixes the issue ?

I see a lot of controller stuff being shuffled around but it's not clear why this helps fixing the issue.

sharidas commented:
@PVince81 
> @sharidas mind adding a summary at the top to explain why this PR fixes the issue ?
> 
> I see a lot of controller stuff being shuffled around but it's not clear why this helps fixing the issue.

I have updated the `Description` and `Motivation and Context` section of the PR. A breif summary has been added to know why this change is added.

PVince81 commented:
@sharidas is this a brand new controller written from scratch or does it reuse some of the code that was moved to the user_management app ?

sharidas commented:
> @sharidas is this a brand new controller written from scratch or does it reuse some of the code that was moved to the user_management app ?

@PVince81 Yes this is a new controller. It is inspired or lets say influenced from user_management's https://github.com/owncloud/user_management/blob/master/lib/Controller/GroupsController.php. Here is the comment, which helped to create this PR https://github.com/owncloud/user_management/pull/96#issuecomment-436304331. Also I have tried to avoid code duplication while writing this controller. Hope this address your query.

PVince81 commented on settings/Controller/GroupsController.php:
```diff
@@ -0,0 +1,124 @@
+			if ($group->getGID() === 'admin') {
+				$adminGroup['id'] = $group->getGID();
+				$adminGroup['name'] = $group->getDisplayName();
+				$adminGroup['userCount'] = 1;
```
I think there can be more than one user in the admin group

PVince81 reviewed (commented):
See comment.
Looks ok otherwise.

sharidas commented:
Current state of the PR

- [x] Replaced admin users count from `1` to `$group->count('')`. This was from the request https://github.com/owncloud/core/pull/33442#discussion_r233865426. I have verified with the latest change, the code does grab all the users from admin or non admin groups.

Hence kindly request for review @PVince81

sharidas commented:
- [x] Updated the test with more than one user in a group.
  - [x] Though there is a test which I updated in this version, I have tried to add another test. In the new test I have not used mocks. It directly alters the db.

@PVince81 I have updated the new version with the unit tests updated. I missed updating the test, in my previous version :(

PVince81 commented:
@sharidas can you click "resolve conversation" on the review items that you solved ?

sharidas commented:
> @sharidas can you click "resolve conversation" on the review items that you solved ?

I have clicked "resolve conversation" on the review items, that were solved with respect to the current state of the PR.

phil-davis commented:
Note: This PR would need to be backported to `stable10` if we some-day release the `user_management` app as a separate app, along with a core 10.* release.
